docs(framework) Add quickstart examples Docker compose guide (#4189)
adap/flower: doc/source/docker/run-quickstart-examples-docker-compose.rst

Run Flower Quickstart Examples with Docker Compose
==================================================

Flower provides a set of `quickstart examples <https://github.com/adap/flower/tree/main/examples>`_
to help you get started with the framework.

> These examples are designed to demonstrate the capabilities of Flower and can be run
> using Docker Compose.

Reviewer: ```suggestion
These examples are designed to demonstrate the capabilities of Flower and by default run using the Simulation Engine. This guide demonstrates how to run them using Flower's Deployment Engine via through Docker Compose.
```

Reviewer: I thought of rewriting it slightly. Probably we can iterate more. Also, should this be a different paragraph from the first sentence in the doc page? or should we join them ?

Author: > I thought of rewriting it slightly. P
That's sounds better! 
I think we can join them. I have no strong preference.

.. important::

   Some quickstart examples may have limitations or requirements that prevent them from running
   on every environment. For more information, please see `Limitations`_.

Prerequisites
-------------

Before you start, make sure that:

- The ``flwr`` CLI is :doc:`installed <../how-to-install-flower>` locally.
- The Docker daemon is running.
- Docker Compose is `installed <https://docs.docker.com/compose/install/>`_.

Run the Quickstart Example
--------------------------

#. Clone the quickstart example you like to run. For example, ``quickstart-sklearn-tabular``:

   .. code-block:: bash

      $ git clone --depth=1 https://github.com/adap/flower.git \

>            && mv flower/examples/quickstart-sklearn-tabular . \
>            && rm -rf flower && cd quickstart-sklearn-tabular

Reviewer: ```suggestion
           && mv flower/examples/quickstart-pytorch . \
           && rm -rf flower && cd quickstart-pythorc
```

Reviewer: i think the pytorch example or the tf would offer a better experience in general

#. Download the `compose.yml <https://github.com/adap/flower/blob/main/src/docker/complete/compose.yml>`_ file into the example directory:

   .. code-block:: bash

      $ curl https://raw.githubusercontent.com/adap/flower/refs/heads/main/src/docker/complete/compose.yml \
          -o compose.yml

#. Build and start the services using the following command:

   .. code-block:: bash

      $ docker compose up --build -d

#. Append the following lines to the end of the ``pyproject.toml`` file and save it:

   .. code-block:: toml
      :caption: pyproject.toml

>       [tool.flwr.federations.docker-compose]

Reviewer: I'm thinking if we should name the federation to something different that doesn't use the `docker` or `	compose` keywords. The alternative would be to add a comment saying that the string that comes after `federations.` is up to the user to define (then use is accordingly when doing `flwr run .`

Author: I'm open to alternative name suggestions. we can also call it `local` or `local-insecure` 

I'll add a comment, but this i feel this information belongs more to the flwr cli docs.

Reviewer: how about `local-deployment`? I agree info should better be left to other parts of the docs.

Author: `local-deployment` sounds good to me!

>       address  =  "127.0.0.1:9093"
>       insecure  = true

Reviewer: ```suggestion
      address = "127.0.0.1:9093"
      insecure = true
```

Run Flower Quickstart Examples with Docker Compose
==================================================

Flower provides a set of `quickstart examples <https://github.com/adap/flower/tree/main/examples>`_
to help you get started with the framework. These examples are designed to demonstrate the
capabilities of Flower and by default run using the Simulation Engine. This guide demonstrates

> how to run them using Flower's Deployment Engine via through Docker Compose.

Reviewer 2: ```suggestion
how to run them using Flower's Deployment Engine via Docker Compose.
```

#. Clone the Flower Repository:

   .. code-block:: bash

      $ git clone --depth=1 https://github.com/adap/flower.git

#. Change the directory to the location of the Docker Compose file:

   .. code-block:: bash

      $ cd flower/src/docker/complete/

#. Set the ``PROJECT_DIR`` environment variable to the path of the quickstart example you want to
   run. The path should be relative to the location of the Docker Compose files:

   .. code-block:: bash

      $ export PROJECT_DIR=../../../examples/quickstart-sklearn-tabular

#. Build and start the services using the following command:

   .. code-block:: bash

      $ docker compose -f compose.yml up --build -d

#. Follow the logs of the SuperExec service:

   .. code-block:: bash

      $ docker compose logs superexec -f

Step 2: Run the Quickstart Example (Terminal 2)
-----------------------------------------------

#. Change the directory to the quickstart example:

   .. code-block:: bash

      $ cd flower/examples/quickstart-sklearn-tabular

#. Append the following lines to the end of the ``pyproject.toml`` file and save it:

   .. code-block:: toml
      :caption: pyproject.toml

>       [tool.flwr.federations.docker-compose]

Author: We could add this to the `pyproject.toml` file to remove this step. wdyt?

Reviewer: I think leaving it for people to do could be a simple but useful learning experience for future projects?

Author: I think it is ok to do it for a single example but i personally got a little bit annoyed when i tried different examples.

      $ docker compose logs superexec -f

#. Switch back to the second terminal and change the directory to the new quickstart
   example you like to run:

   .. code-block:: bash
      :caption: Terminal 2

      $ cd ../quickstart-pytorch

#. Append the following lines to the end of the ``pyproject.toml`` file and save it:

   .. code-block:: toml
      :caption: pyproject.toml

      [tool.flwr.federations.docker-compose]
      address  =  "127.0.0.1:9093"
      insecure  = true

#. Run the example:

   .. code-block:: bash
      :caption: Terminal 2

      $ flwr run . docker-compose

Clean Up
--------

Remove all services and volumes:

.. code-block:: bash

   $ docker compose down -v

Limitation
----------

.. list-table::
   :header-rows: 1

   * - Quickstart Example
     - Limitation/Requirements
   * - quickstart-fastai
     - None
   * - examples/quickstart-huggingface
     - For CPU-only environments, it requires at least 32GB of memory.
   * - quickstart-jax
     - The example has not yet been updated to work with the latest ``flwr`` version.
   * - quickstart-mlcube
     - The example has not yet been updated to work with the latest ``flwr`` version.

>    * - quickstart-mlx

Author: We should remove the mentioning of Docker here: https://github.com/adap/flower/blob/541064b79e4de3d95be9ec270ff66da201feb3b6/examples/quickstart-mlx/README.md?plain=1#L70

Reviewer: https://github.com/adap/flower/pull/4202

     - `Requires to run on macOS with Apple Silicon <https://ml-explore.github.io/mlx/build/html/install.html#python-installation>`_.
   * - quickstart-monai
     - None
   * - quickstart-pandas
     - The example has not yet been updated to work with the latest ``flwr`` version.
   * - quickstart-pytorch-lightning

>      - Requires an older pip version that is not supported by the Flower Docker images.

Author: If we update the dependencies (pytorch) we can make the quickstart work with docker compose
